### PR TITLE
Update cert pinning logic.

### PR DESCRIPTION
**Pull request by chogorma**

### CODE REVIEW

File: Release/include/cpprest/certificate_info.h [hunk 1]
```diff
@@ -0,0 +1,47 @@
+/***
+* ==++==
+*
+* Copyright (c) Microsoft Corporation. All rights reserved.
```
Comment (coliemcgarry) [inline Release/include/cpprest/certificate_info.h:4]: Is this really a MS header, or something you wrote?

File: Release/src/http/client/http_client_asio.cpp [hunk 2]
```diff
@@ -1459,14 +1462,11 @@ class asio_context : public request_context, public std::enable_shared_from_this
...
 
-            if (!verify_cert_chain_platform_specific(verifyCtx, utility::conversions::to_utf8string(host)))
-            {
-                return false;
-            }
-            else
-            {
-                return m_http_client->client_config().invoke_certificate_chain_callback(host, get_X509_cert_chain_encoded_data(verifyCtx));
-            }
+            auto chainFunc = [this](const std::shared_ptr<certificate_info>& cert_info) {
```
Comment (coliemcgarry) [inline Release/src/http/client/http_client_asio.cpp:1465]: Are we ok with the lifecycle of `this`?

File: Release/src/http/client/http_client_winhttp.cpp [hunk 3]
```diff
@@ -1085,32 +1086,45 @@ class winhttp_client : public _http_client_communicator
...
+                        p_request_context->m_certificate_chain_verification_failed = false;
+                    }
+                    else
+                    {
+                        p_request_context->m_certificate_chain_verification_failed = true;
+                    }
+                }
+                else
                 {
                     p_request_context->m_certificate_chain_verification_failed = true;
```
Comment (coliemcgarry) [inline Release/src/http/client/http_client_winhttp.cpp:1129]: I know this line is not changed, but if we don't have a callback, then is the verification_failed result not related to the info->verified result?

File: Release/src/http/client/x509_cert_utilities_apple.cpp [hunk 4]
```diff
@@ -99,28 +99,58 @@ bool verify_X509_cert_chain(const std::vector<std::string> &certChain, const std
...
-    }
 
-    return false;
+        if (certInfoFunc)
+        {
+            auto info = std::make_shared<certificate_info>(hostName);
+            info->certificate_error = (long)trustResult;
+            info->verified = isVerified;
+
+            CFIndex cnt = SecTrustGetCertificateCount(trust.get());
```
Comment (coliemcgarry) [inline Release/src/http/client/x509_cert_utilities_apple.cpp]: You could extract the following into a helper method to 'build certificate info'. 

We're adding quite a bit to this method

File: Release/src/http/client/x509_cert_utilities_apple.cpp [hunk 5]
```diff
@@ -99,28 +99,58 @@ bool verify_X509_cert_chain(const std::vector<std::string> &certChain, const std
...
+                full_cert_chain.reserve(cnt);
+                for (int i = 0; i < cnt; i++)
+                {
+                    SecCertificateRef cert = SecTrustGetCertificateAtIndex(trust.get(), i);
+                    cf_ref<CFDataRef> cdata = SecCertificateCopyData(cert);
+
+                    const unsigned char * buffer = CFDataGetBytePtr(cdata.get());
+                    full_cert_chain.emplace_back(std::vector<unsigned char>(buffer, buffer + CFDataGetLength(cdata.get())));
+                }
+                info->certificate_chain = full_cert_chain;
```
Comment (coliemcgarry) [inline Release/src/http/client/x509_cert_utilities_apple.cpp]: Why copy?, why not create this inplace?

File: Release/include/cpprest/certificate_info.h (hunk 1 above)
Comment (chogorma) [reply Release/include/cpprest/certificate_info.h:4]: If it's going back to the MS repo it needs to have that, that if we ever get around to open a PR with them.

File: Release/src/http/client/http_client_asio.cpp (hunk 2 above)
Comment (chogorma) [reply Release/src/http/client/http_client_asio.cpp:1465]: That's what was happening before, i just changed it back to what we where doing before.
I think the lifecycle is good it's executed in a synchronous call.

File: Release/src/http/client/x509_cert_utilities_apple.cpp (hunk 5 above)
Comment (chogorma) [reply Release/src/http/client/x509_cert_utilities_apple.cpp]: Good point will change that.

File: Release/src/http/client/http_client_winhttp.cpp (hunk 3 above)
Comment (coliemcgarry) [reply Release/src/http/client/http_client_winhttp.cpp:1129]: Got it - had the logic inverted!

File: Release/src/http/client/x509_cert_utilities_win32.cpp [hunk 6]
```diff
@@ -76,54 +91,66 @@ bool verify_X509_cert_chain(const std::vector<std::string> &certChain, const std
...
+            chain->TrustStatus.dwErrorStatus == CERT_TRUST_REVOCATION_STATUS_UNKNOWN ||
+            chain->TrustStatus.dwErrorStatus == (CERT_TRUST_IS_OFFLINE_REVOCATION | CERT_TRUST_REVOCATION_STATUS_UNKNOWN))
+        {
+            isVerified = true;
+        }
 
-    chain.reset(chainContext);
+        if (certInfoFunc)
+        {
+            auto info = std::make_shared<certificate_info>(hostName);
```
Comment (coliemcgarry) [inline Release/src/http/client/x509_cert_utilities_win32.cpp]: Maybe break out into a buildCertInfo()

**Pull request merged**